src/lib.rs

```diff
@@ -595,12 +642,87 @@ impl Cancelled {
 
 impl std::fmt::Display for Cancelled {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.write_str("cancelled")
+        let why = match self {
+            Cancelled::PendingWrite => "pending write",
+            Cancelled::PropagatedPanic => "propagated panic",
+            Cancelled::UnexpectedCycle(_) => "unexpected cycle",
+        };
+        f.write_str("cancelled because of ")?;
+        f.write_str(why)
     }
 }
 
 impl std::error::Error for Cancelled {}
 
+/// Information about an "unexpected" cycle, meaning one where some of the
+/// participants lacked cycle recovery annotations.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Cycle {
+    participants: plumbing::CycleParticipants,
+}
+
+impl Cycle {
+    pub(crate) fn new(participants: plumbing::CycleParticipants) -> Self {
+        Self { participants }
+    }
+
+    /// Iterate over the [`DatabaseKeyIndex`] for each query participating
+    /// in the cycle. The start point of this iteration within the cycle
+    /// is arbitrary but deterministic, but the ordering is otherwise determined
+    /// by the execution.
+    pub fn participant_keys(&self) -> impl Iterator<Item = DatabaseKeyIndex> + '_ {
+        let min = self.participants.iter().min().unwrap();
```

Comment:
Can we rotate `participants` in `Cycle::new` instead, so that they are always just in the right order?

Reply:
We could. Cute idea.